### PR TITLE
many: add new "install-mode: disable" option

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds a new options to disable startup of daemons. The
default of snapd is that all daemons are started once you install
the snap. However that is sometimes not desired and our users
need to write fragile wrapper scripts to workaround the missing
native support.

With the new `install-mode: disable` the service is not started
on install and can be enabled via snapctl from a hook or via a
management agent.

This is a first RFC to get feedback on naming and if we can get
away with a simple implementation like this. Unit tests missing
but will be added once there is more direction.

### CODE REVIEW

**Comment by zyga** [comment]
@mvo5 I'm very happy to see this but I would love to frame this in the existing language of presets. Perhaps a service should be started on a desktop but not on a server or IOT device. If we had language to just offer a default preset that is identical in expressiveness to this proposal, we could eventually grow it without drifting away from the basic capability.

File: overlord/snapstate/backend/link.go
Comment (stolowski) [inline overlord/snapstate/backend/link.go:193]: This will conflict with changed semantics in #8892 ; after my changes to wrappers and snapstate land all the units will be created unconditionally (but not enabled) on link, and will get enabled later on start-snap-services. At this point I'd prefer for 8892 to land first and then adjust this code - it will need to be executed from startSnapServices I suppose, but there is no linkCtx there so other adjustments will be required.

Comment (pedronis) [reply overlord/snapstate/backend/link.go:193]: I agree with Pawel on the ordering here. About the name, something parallel to refresh-mode would be best, install-mode is tempting but we'll see.

Comment (mvo5) [reply overlord/snapstate/backend/link.go:193]: Happy to wait for 8892 to land first and tweak then

Comment (stolowski) [reply overlord/snapstate/backend/link.go:193]: The PR to wait for is now https://github.com/snapcore/snapd/pull/8960

**Comment by mvo5** [comment]
I updated the name of the option to `install-mode: daemon-inhibit` to make this a bit more clear and similar to refresh-mode. 

My only small concern is that if there is ever something orthogonal that we may want to add to install-mode (I can't think of anything right now) we need to think what to do.

File: overlord/snapstate/backend/link.go
Comment (pedronis) [inline overlord/snapstate/backend/link.go:196]: I think we need to decide on this, my thinking was that the decision shouldn't be based whether we are installing the snap itself, but whether the service exists atm or not yet, that's the "installl" of the service, not the snap. We might need to think whether that impacts refresh-mode behavior or not.

File: overlord/snapstate/backend/link.go
Comment (pedronis) [inline overlord/snapstate/backend/link.go:200]: my thinking was that to follow "refresh-mode" and what we control do, this should be "disable"

**Review by pedronis — COMMENTED** [review]
some considerations on the draft. As I mentioned elsewhere if this is for 2.49, not 2.48, it should probably be done on top of #8960

File: overlord/snapstate/backend/link.go
Comment (mvo5) [reply overlord/snapstate/backend/link.go:197]: Agreed, I think this code is wrong, I added a XXX here and in the spread test how this should work so that upgrades also work well.

File: overlord/snapstate/backend/link.go
Comment (mvo5) [reply overlord/snapstate/backend/link.go]: This is now based on 8960

File: overlord/snapstate/backend/link.go
Comment (mvo5) [reply overlord/snapstate/backend/link.go]: The code now looks at the previous installed snap and only applies "install-mode: disable" to new services.

File: snap/info_snap_yaml.go
Comment (pedronis) [inline snap/info_snap_yaml.go:368]: we need validation code in ValidateApp like for RefreshMode,  we should allow for "enable" and "disable", where "enable" is the default

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:1723]: I found this comment slightly confusing, how about something like:
```
// installModeDisabledServices returns what services with
// "install-mode: disabled" should be disabled. Only services
// seen for the first time are considered.
```
?

**Review by pedronis — COMMENTED** [review]
thanks, main logic looks right to me. need validation code though and @stolowski review

also seems the spread test needs more work?

File: tests/main/snap-service/task.yaml
Comment (stolowski) [inline tests/main/snap-service/task.yaml:53]: We will also need a test where service gets enabled via hook.

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:1734]: I spent a good while pondering about why this is needed (and played around with removing this logic & running the spread test); this actually handles the case where we install the given service for the first time (which could be a refresh to a new revision), and otherwise we start+enable it if it got enabled by other means (such as `snap enable..`). Perhaps worth a comment.

**Review by stolowski — COMMENTED** [review]
Thanks, I think this looks fine! (but yes, needs a few more tests). Added some small remarks regarding comments. 

I'm also wondering if `install-mode: disable` is the best term to express the semantics. What really happens is we do not start/enable the service (effectively ignore the service on install, leaving starting/enabling it to the user or a hook); 'disable' may suggest it remains disabled after install no matter what, while it could be started by hooks. Perhaps `install-mode: manual' or '..: ignore' ?

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:1784]: I think this needs to take into account also ServicesEnabledByHooks?  is that right @stolowski ?

We need a test where a post-refresh or an install hook enables a service that is install-mode: disable maybe?

**Review by pedronis — COMMENTED** [review]
one open question

**Comment by pedronis** [comment]
@stolowski 
> I'm also wondering if install-mode: disable is the best term to express the semantics. What really happens is we do not start/enable the service (effectively ignore the service on install, leaving starting/enabling it to the user or a hook); 'disable' may suggest it remains disabled after install no matter what, while it could be started by hooks. Perhaps `install-mode: manual' or '..: ignore' ?

we are trying to follow if we can the pattern of `refresh-mode: endure|restart`

the values need to be verbs and  if possible also connect to the operations/actions we have on services (like restart). Here as well "endure" doesn't stop a hook from restarting the service anyway...

A different approach would be to not stick to the verb pattern but have maybe `initially-disabled`, `initially-enabled`

File: overlord/snapstate/handlers.go
Comment (stolowski) [reply overlord/snapstate/handlers.go:1784]: Yes, you're right, I think the spread test I suggested earlier below would uncover this.

File: overlord/snapstate/handlers.go
Comment (mvo5) [reply overlord/snapstate/handlers.go:1734]: I added/updated the comment (and also picked the comment you suggested above.

File: tests/main/snap-service/task.yaml
Comment (mvo5) [reply tests/main/snap-service/task.yaml:53]: Thanks, I will look into this

File: overlord/snapstate/handlers.go
Comment (mvo5) [reply overlord/snapstate/handlers.go:1792]: I added a unit test and code to handle this but the spread test for this is not done yet

Comment (mvo5) [reply overlord/snapstate/handlers.go:1792]: I also added a spread test now with a `snapctl start --enable` from the install hook.

File: snap/info_snap_yaml.go
Comment (mvo5) [reply snap/info_snap_yaml.go:368]: This is added now.

File: snap/validate.go
Comment (pedronis) [inline snap/validate.go:788]: as I wrote I think we should also support "enable" which would be the default (same as "")

**Review by pedronis — APPROVED** [review]
thanks, one consideration, I haven't looked super closely to the spread test

File: overlord/snapstate/handlers.go
Comment (stolowski) [inline overlord/snapstate/handlers.go:1743]: Thanks for this comment!

File: tests/main/snap-service-install-mode/task.yaml
Comment (stolowski) [inline tests/main/snap-service-install-mode/task.yaml:11]: Thanks for this!

**Review by stolowski — APPROVED** [review]
Nice, thanks for the changes! +1

**Pull request merged**
